core/storage/trie/polkadot_trie_db/polkadot_trie_batch.cpp

```diff
@@ -0,0 +1,107 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "storage/trie/polkadot_trie_db/polkadot_trie_batch.hpp"
+
+namespace kagome::storage::trie {
+
+  using common::Buffer;
+
+  PolkadotTrieBatch::PolkadotTrieBatch(PolkadotTrieDb &trie) : trie_{trie} {}
+
+  outcome::result<void> PolkadotTrieBatch::put(const Buffer &key,
+                                               const Buffer &value) {
+    auto value_copy = value;
+    return put(key, std::move(value_copy));
+  }
+
+  outcome::result<void> PolkadotTrieBatch::put(const Buffer &key,
+                                               Buffer &&value) {
+    if (value.empty()) {
+      OUTCOME_TRY(remove(key));
+    } else {
+      commands_.push_back({Action::PUT, key, std::move(value)});
+    }
+    return outcome::success();
+  }
+
+  outcome::result<void> PolkadotTrieBatch::remove(const Buffer &key) {
+    commands_.push_back({Action::REMOVE, key});
+    return outcome::success();
+  }
+
+  outcome::result<void> PolkadotTrieBatch::commit() {
+    if (commands_.empty()) {
+      return outcome::success();
+    }
+
+    // make the command list empty and store commands in a local list
+    // this is done because if an error occurs, commands that are not processed
+    // yet must not remain in the batch
+    decltype(commands_) commands{};
```

Comment:
auto commands_ = std::move(commands); // doesn't it work?

Reply:
Perhaps it does, but I need exactly to make commands_ empty and commands possess the content of commands_, and I'm not sure if 'commands = std::move(commands_)' is certain to do this, or it will leave commands_ in an invalid state.

Reply:
"Unless otherwise specified, all standard library objects that have been moved from are placed in a valid but unspecified state. "